lib/AST/Type.cpp

```diff
@@ -556,19 +556,18 @@ bool TypeBase::hasTypeRepr() const {
   });
 }
 
-bool TypeBase::isVoid() {
-  if (auto TT = getAs<TupleType>())
-    return TT->getNumElements() == 0;
-  return false;
-}
-
-/// Check if this type is equal to Swift.Bool.
-bool TypeBase::isBool() {
-  if (auto NTD = getAnyNominal())
-    if (isa<StructDecl>(NTD))
-      return getASTContext().getBoolDecl() == NTD;
-  return false;
-}
+#define KNOWN_STDLIB_TYPE_DECL(NAME, DECL_CLASS, NUM_GENERIC_PARAMS) \
+bool TypeBase::is##NAME() { \
+  if (auto generic = getAnyGeneric()) \
+    if (isa<DECL_CLASS>(generic)) \
+      return getASTContext().get##NAME##Decl() == generic; \
+  /** Edge case for Void, could be () too */ \
+  if (#NAME == "Void") \
```

Comment:
You're doing a pointer comparison on the string literals, which isn't really guaranteed to work. Convert one of them to StringRef (or llvm::StringLiteral) to get the behavior you want.

(And in practice, you might as well do this first and have the other part in an `else` branch.)

Reply:
Also, the way this logic is set up, you're still never going to get a TypeAliasDecl to return `true` here (which is probably a good thing for types). Maybe it'd be better to just separate typealiases out specially on TypeBase, and then `isVoid` can go back to being manually implemented.

Reply:
I want to clean up KnownStdlibTypes soon, so I can do things like separate typealiases out, but I'll save those cleanups for a future patch. In any case, I'd like it if you reviewed the new implementation of this.